## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [94]:
import pandas as pd
client_df = pd.read_csv('Resources/client_dataset.csv')
client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [95]:
# View the column names in the data
client_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [96]:
# Use the describe function to gather some basic statistics
client_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [98]:
# Use this space to do any additional research
# and familiarize yourself with the data.

#Get number of rows and columns
client_df.shape

#Check datatypes of each column
client_df.dtypes

#check how many nulls are there
client_df.isnull().sum()

#check for any duplicate rows
client_df.duplicated()

# drop any duplicates
client_df.drop_duplicates()

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
0                1  
1                0  
2                6  
3                3  
4                1  
...            ...  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5  

[54639 rows x 18 columns]

In [99]:
# What three item categories had the most entries?
client_df['category'].value_counts()[:3]


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [100]:
# For the category with the most entries,
# which subcategory had the most entries?
consumables_client_df = client_df[(client_df['category'] == 'consumables')]
consumables_client_df['subcategory'].value_counts()[:1]

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [101]:
# Which five clients had the most entries in the data?
client_df['client_id'].value_counts()[:5]

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [102]:
# Store the client ids of those top 5 clients in a list.
top_clients = client_df['client_id'].value_counts().nlargest(5).index.tolist()
top_clients


[33615, 66037, 46820, 38378, 24741]

In [103]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_df['qty'].loc[(client_df['client_id'] == top_clients[0])].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [104]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
client_df['sub_total'] = (client_df['qty'] * client_df['unit_price'])
client_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  sub_total  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [105]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def get_shipping_price(x):
    """
        This function is to calculate shipping price.

        Optional:
        use unit_weight, qty to calculate the weight and then 
        shipping price.

        Args:
        Row of the dataframe

        Returns:
        Float value which is calculated shipping price
    """
    total_weight = (x['unit_weight'] * x['qty'])
    
    if total_weight <= 50:
        return round((total_weight * 10),2)
    return round((total_weight * 7),2)

client_df['shipping_price'] = client_df.apply(get_shipping_price, axis=1)
client_df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  sub_total  shipping_price  
0                1  115164.00         5512.50  
1                0     523.95          312.90  
2                6     527.28          458.64  
3                3    1056.18          356.70  
4                1    3902.00         6500.20  
...            ...        ...             ...  
54634            8    2743.29          519.75  
54635            0    9709.73         3849.30  
54636            6   31188.50        13832.00  
54637            0     165.76        14143.36  
54638            5    3

In [106]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
client_df['total_price'] = round(((client_df['sub_total'] + \
                                   client_df['shipping_price']) * 1.0925), 2)
client_df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number  sub_total  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       shipping_price  total_price  
0             5512.50    131839.08  
1              312.90       914.26  
2              458.64      1077.12  
3              356.70      1543.57  
4             6500.20     11364.40  
...               ...          ...  
54634          519.75      3564.87  
54635         3849.30     14813.24  
54636        13832.00     49184.90  
54637        14143.36     15632.71  


In [107]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

#it can also be done using function. 
""" def calculate_cost(x):
    return ((x['unit_cost'] * x['qty']) + x['shipping_price'])

client_df['total_cost'] = client_df.apply(calculate_cost, axis=1) """

#can also be implemented using lamda
client_df['total_cost'] = client_df.\
    apply(lambda x: ((x['unit_cost'] * x['qty']) + x['shipping_price']), axis=1)
client_df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number  sub_total  shipping_price  total_price  \
0             7.50   105            1  115164.00         5512.50    131839.08   
1             1.49    21            0     523.95          312.90       914.26   
2             1.68    39            6     527.28          458.64      1077.12   
3             1.23    29            3    1056.18          356.70      1543.57   
4            46.43    20            1    3902.00         6500.20     11364.40   
...            ...   ...          ...        ...             ...          ...   
54634         2.25    33            8    2743.29          519.75      3564.87   
54635        11.70    47            0    9709.73         3849.30     14813.24   
54636         4.16   475            6   31188.50        13832.00     49184.90   
54637        18.04   112            0     165.76        14143.36     15632.71   
54638         2.07  1031            5    3103.31        14939.19     19711.43   

       total_cost  
0        85597.05  
1          629.79  
2          765.18  
3         1077.35  
4         8663.60  
...           ...  
54634     2222.55  
54635    12095.92  
54636    41054.25  
54637    14280.00  
54638  

In [108]:
# Create a column for the profit of
# each line using line cost and line price

client_df['profit'] = client_df.apply(lambda i: (i['total_price'] - \
                                                 i['total_cost']), axis=1)
client_df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  sub_total  shipping_price  total_price  total_cost  \
0                1  115164.00         5512.50    131839.08    85597.05   
1                0     523.95          312.90       914.26      629.79   
2                6     527.28          458.64      1077.12      765.18   
3                3    1056.18          356.70      1543.57     1077.35   
4                1    3902.00         6500.20     11364.40     8663.60   
...            ...        ...             ...          ...         ...   
54634            8    2743.29          519.75      3564.87     2222.55   
54635            0    9709.73         3849.30     14813.24    12095.92   
54636            6   31188.50        13832.00     49184.90    41054.25   
54637            0     165.76        14143.36     15632.71    14280.00   
54638            5    3103.31        14939.19     19711.43    16599.10   

         profit  
0      46242.03  
1        284.47  
2        311.94  
3        466.22  
4       2700.80  
...         ...  
54634   1342.32  
54635   2717.32  
54636   8130.65  
54637   1352.71  
54638   3112.33  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [109]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]

for i in order_ids:
  print(f"Order ID: {i}, Total Price: ${client_df['total_price'].loc[(client_df['order_id'] == i)].sum():,.2f}")


Order ID: 2742071, Total Price: $152,811.89
Order ID: 2173913, Total Price: $162,388.71
Order ID: 6128929, Total Price: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [110]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_spenders = {}

# Sort the DataFrame by total_price in descending order and get top 5 client_ids
for i in top_clients:
    top_spenders[i] = round((client_df['total_price']\
                             .loc[(client_df['client_id'] == i)].sum()), 2)
    
print("Top 5 Client IDs:", top_spenders)

Top 5 Client IDs: {33615: 8377308.52, 66037: 10259514.79, 46820: 9743794.36, 38378: 12906550.87, 24741: 82268892.02}


In [111]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Extract the rows with just top client ids
top_5_clients_df = client_df.loc[(client_df['client_id'].isin(top_clients))]
top_5_clients_df

# create lists of data for each column you want in summary
top_5_clients = []
top_5_total_units = []
top_5_total_shipping_price = []
top_5_total_revenue = []
top_5_total_profits = []

# populate the lists in same order of client id so you can make sure 
# the data is correct
for i in top_clients:
    top_5_clients.append(i)
    top_5_total_units.append(client_df['qty']\
                             .loc[(client_df['client_id'] == i)].sum())
    top_5_total_shipping_price.append(round((client_df['shipping_price']\
                             .loc[(client_df['client_id'] == i)].sum()), 2))
    top_5_total_revenue.append(top_spenders[i])
    top_5_total_profits.append(round((client_df['profit']\
                             .loc[(client_df['client_id'] == i)].sum()), 2))

# use above lists to create the summary dataframe
top_clients_summary_df = \
    pd.DataFrame({'client_id': top_5_clients, \
                    'total_units': top_5_total_units, \
                    'total_shipping': top_5_total_shipping_price, \
                    'total_revenue': top_5_total_revenue, \
                    'total_profit': top_5_total_profits
                }).sort_values(by='total_profit')

#top_clients_summary_df.dtypes
top_clients_summary_df

client_id  total_units  total_shipping  total_revenue  total_profit
0      33615        64313      1828984.89     8377308.52    2201994.61
2      46820        75768      1601448.84     9743794.36    2736602.72
1      66037        43018      1395151.85    10259514.79    3255031.81
3      38378        73667      3429455.40    12906550.87    3271829.89
4      24741       239862      5126448.37    82268892.02   36579992.31

In [112]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# convert all currency values to millions
top_clients_summary_df[['total_shipping', 'total_revenue', 'total_profit']] = \
    top_clients_summary_df[['total_shipping', 'total_revenue', 'total_profit']] \
        .apply(lambda x: round(x/1000000,2))

# rename the columns
top_clients_summary_df.rename(\
    columns={'client_id': 'Client ID', \
                'total_units': 'Total Units', \
                'total_shipping': 'Total Shipping (in Millions)', \
                'total_revenue': 'Total Revenue (in Millions)', \
                'total_profit': 'Total Profit (in Millions)'}, inplace=True)


top_clients_summary_df


Client ID  Total Units  Total Shipping (in Millions)  \
0      33615        64313                          1.83   
2      46820        75768                          1.60   
1      66037        43018                          1.40   
3      38378        73667                          3.43   
4      24741       239862                          5.13   

   Total Revenue (in Millions)  Total Profit (in Millions)  
0                         8.38                        2.20  
2                         9.74                        2.74  
1                        10.26                        3.26  
3                        12.91                        3.27  
4                        82.27                       36.58

In [120]:
# Sort the updated data by "Total Profit" form highest to lowest
top_clients_summary_sorted_df = top_clients_summary_df.sort_values('Total Profit (in Millions)', ascending=False)
top_clients_summary_sorted_df.reset_index(drop=True, inplace=True)
top_clients_summary_sorted_df

# write the updated and sorted dataframe to csv
#top_clients_summary_sorted_df.to_csv("Output/top_clients_summary.csv")

Client ID  Total Units  Total Shipping (in Millions)  \
0      24741       239862                          5.13   
1      38378        73667                          3.43   
2      66037        43018                          1.40   
3      46820        75768                          1.60   
4      33615        64313                          1.83   

   Total Revenue (in Millions)  Total Profit (in Millions)  
0                        82.27                       36.58  
1                        12.91                        3.27  
2                        10.26                        3.26  
3                         9.74                        2.74  
4                         8.38                        2.20

In [ ]:
''' 
    Above data helps us to identify how much profit company made with 
    top 5 clients. We can understand how much was spent on shipping and see if 
    there is a way to reduce it based on units shipped. We can determine 
    if any additional discounts or incentives can be given to any of these top 
    clients based on the revenue they generated.

'''